### PR TITLE
AsyncState API Review

### PR DESCRIPTION
**Pull request by geeknoid**
The only thing to review is the definition of the IAsyncLocalContext type.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4467)

### CODE REVIEW

**Review by rafal-mz — APPROVED** [review]

File: API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace Microsoft.Extensions.AsyncState;
+
+/// <summary>
+/// Provides access to the current async context stored outside of the HTTP pipeline.
+/// </summary>
+/// <typeparam name="T">The type of the asynchronous state.</typeparam>
+/// <remarks>This type is intended for internal use. Use <see cref="T:Microsoft.Extensions.AsyncState.IAsyncContext`1" /> instead.</remarks>
+[Experimental("EXTEXP0006", UrlFormat = "https://aka.ms/dotnet-extensions-warnings/{0}")]
+[EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (xakep139) [inline API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs:14]: Why do we need it to be public?

File: API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+// Assembly 'Microsoft.Extensions.AsyncState'
+
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.Extensions.AsyncState;
+
+/// <summary>
+/// Provides access to the current async context stored outside of the HTTP pipeline.
```
Comment (xakep139) [inline API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs:9]: Should we really mention `HTTP` here? The package is agnostic

File: API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs (hunk 1 above)
Comment (geeknoid) [reply API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs:14]: Cause it's used from Microsoft.AspNetCore.AsyncState.

File: API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs (hunk 2 above)
Comment (geeknoid) [reply API/WithDocs/Microsoft.Extensions.AsyncState/IAsyncLocalContext.cs:9]: Good point, I'll remove.

**Pull request closed without merging**